### ISSUE TITLE
nn.Module.forward signature with **kwargs

### ISSUE THREAD

**Issue by prokotg** [pytorch/pytorch#23732]
Hi there!

The other day I was working on improving `utils.checkpoint` module and I found out there is no way right now to pass `**kwargs `to checkpointed function. This is probably in order to by compliant with the `nn.Module.forward` method. 

Now, I could not think about any argument against having `**kwargs` in `forward` method and many people overload this method with `**kwargs` because it is nice and lit :) I was about to change it and create pull request but before I do it I decided to go smart and ask you guys if there is some obvious reason why we should not do it. Otherwise, I am more than happy to start working on it.

Thanks :- )

**Comment by fmassa** [comment]
Adding a `**kwargs` argument to the base `forward` method won't change anything in the functionality I believe, given that `__call__` already has thr kwargs and passes it to `forward`, see https://github.com/pytorch/pytorch/blob/b2f6e2bdc1373e799f20adaa97ea287a06f03578/torch/nn/modules/module.py#L529

**Comment by SsnL** [comment]
There is nothing that prevents it but you gain nothing from this change as well because forward is always overriden.

**Comment by prokotg** [comment]
True, no gain in functionality, just a change to be consistent I guess. I am going to try to change `checkpoint` module to take into account `**kwargs` though. 

Thank you :)

**Comment by saareliad** [comment]
Currently it could be solved if programmers write `forward` functions in a way which `checkpoint` is able to pass `*args` to it (in order) instead using kwargs.

**OR** 

Pass a dict `function_kwargs` to `checkpoint`, which will end up calling:

`CheckpointFunction.apply(function, preserve, *args, **function_kwargs)`
(and change `CheckpointFunction` accordingly)
instead of 
`CheckpointFunction.apply(function, preserve, *args)`

For example, `function_kwargs` could be added a kwarg to the `checkpoint` function.

IMO should be easy to implement in a backward compatible way.

**Comment by prokotg** [comment]
@saareliad right, that's what the documentation says. The ```forward``` function should know what to do with `**args` so we have to pass it in some order and indeed you have to know what to do with input tuple. However, your ```forward``` method might require, for example, some parameters and you can have multiple of them and passing them without keyword could get messy. 

Besides, current solution forces developers to rewrite their current, key-worded model to be keyword-args-less and I think since pytorch is aiming to be python-first, this should not take place.

I followed the data-flow path and basically started to change some code to incorporate kwargs
here
https://github.com/pytorch/pytorch/blob/b2f6e2bdc1373e799f20adaa97ea287a06f03578/torch/csrc/autograd/python_function.cpp#L1056

and in the function itself to unpack the `**kwargs` dict

https://github.com/pytorch/pytorch/blob/b2f6e2bdc1373e799f20adaa97ea287a06f03578/torch/csrc/autograd/python_function.cpp#L675

and I stumbled upon #16940
